fs/zbd_zenfs.cc

```diff
@@ -711,6 +717,55 @@ IOStatus ZonedBlockDevice::AllocateEmptyZone(Zone **zone_out) {
   return IOStatus::OK();
 }
 
+int ZonedBlockDevice::DirectRead(char *buf, uint64_t offset, uint32_t n) {
+  int ret = 0;
+  int f = GetReadDirectFD();
+
+  while (ret < n) {
+    int r = pread(f, buf, n, offset);
+    if (r == -1) {
+      return -1;
+    }
+    if (r == 0) {
+      break;
+    }
+    ret += r;
+  }
+
+  return ret;
+}
+
+IOStatus ZonedBlockDevice::ReleaseMigrateZone(Zone *zone) {
+  {
+    std::unique_lock<std::mutex> lock_(migrate_zone_mtx_);
+    migrating_ = false;
+    if (zone != nullptr) {
+      zone->CheckRelease();
+      Info(logger_, "ReleaseMigrateZone: %lu", zone->start_);
+    }
+  }
+  migrate_resource_.notify_one();
+  return IOStatus::OK();
+}
+
+IOStatus ZonedBlockDevice::TakeMigrateZone(Zone **out_zone,
+                                           Env::WriteLifeTimeHint file_lifetime,
+                                           uint32_t min_capacity) {
+  std::unique_lock<std::mutex> lock_(migrate_zone_mtx_);
+  migrate_resource_.wait(lock_, [this] { return !migrating_; });
+
+  migrating_ = true;
+
+  unsigned int best_diff = LIFETIME_DIFF_NOT_GOOD;
+  auto s =
+      GetBestOpenZoneMatch(file_lifetime, &best_diff, out_zone, min_capacity);
```

Comment:
My first thought was to fill a half-written zone could reduce data fragmentation, I can test if it helps to use empty zones.

Reply:
It's good to try to fill open zones, but if there is none available for writing, we should move the data to an empty one rather than stopping the migration

Reply:
As agreed over chat, Let's experiment with the current implementation and add migration to empty zones if needed.